### PR TITLE
Update and fix dependencies related to mac install

### PR DESCRIPTION
**Pull request by maxhuettenrauch**
Addresses part of #1015 

### Dependencies

- move jsonargparse and docstring-parser to dependencies to run hl examples without dev
- create mujoco-py extra for legacy mujoco envs
- updated atari extra
    - removed atari-py and gym dependencies
    - added ALE-py, autorom, and shimmy
- created robotics extra for HER-DDPG

### Mac specific

- only install envpool when not on mac
- mujoco-py not working on macOS newer than Monterey (https://github.com/openai/mujoco-py/issues/777)
- D4RL also fails due to dependency on mujoco-py (https://github.com/Farama-Foundation/D4RL/issues/232)

### Other

- reduced training-num/test-num in example files to a number ≤ 20 (examples with 100 led to too many open files)
- rendering for Mujoco envs needs to be fixed on gymnasium side (https://github.com/Farama-Foundation/Gymnasium/issues/749)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -96,6 +96,23 @@ You can also install the dev requirements by adding `--with dev` or the extras
...
+- `mujoco-py` (for legacy mujoco-py environments[^1])
+- `pybullet` (for pybullet environments)
+- `robotics` (for gymnasium-robotics environments)
+- `vizdoom` (for ViZDoom environments)
+- `envpool` (for envpool integration)
+- `dev` (for development requirements)
+
+[^1]: `mujoco-py` is a legacy package and is not recommended for new projects. 
+It is only included for compatibility with older projects. 
+Also note, that there may be compatibility issues with macOS newer than
```
Comment (MischaPanch) [inline README.md]:
```suggestion
Also note that there may be compatibility issues with macOS newer than 
```

File: README.md [hunk 2]
```diff
@@ -96,6 +96,23 @@ You can also install the dev requirements by adding `--with dev` or the extras
...
+- `atari` (for Atari environments)
+- `box2d` (for Box2D environments)
+- `classic_control` (for classic control (discrete) environments)
+- `mujoco` (for MuJoCo environments)
+- `mujoco-py` (for legacy mujoco-py environments[^1])
+- `pybullet` (for pybullet environments)
+- `robotics` (for gymnasium-robotics environments)
+- `vizdoom` (for ViZDoom environments)
+- `envpool` (for envpool integration)
+- `dev` (for development requirements)
```
Comment (MischaPanch) [inline README.md]: Isn't that a group rather than an extra?

File: examples/box2d/bipedal_hardcore_sac.py [hunk 3]
```diff
@@ -60,13 +60,22 @@ def __init__(self, env, action_repeat=3, reward_scale=5, rm_done=True):
     def step(self, action):
         rew_sum = 0.0
         for _ in range(self.action_repeat):
-            obs, rew, done, info = self.env.step(action)
+            step_result = self.env.step(action)
+            if len(step_result) == 4:
```
Comment (MischaPanch) [inline examples/box2d/bipedal_hardcore_sac.py]: This should never happen, we use gymnasium envs, not (older) gym envs. How did you run into this?

Comment (MischaPanch) [reply examples/box2d/bipedal_hardcore_sac.py]: Note that if it needs fixing in one example, it will need fixing in all of them :(

File: examples/mujoco/fetch_her_ddpg.py [hunk 4]
```diff
@@ -30,7 +29,7 @@
 
 def get_args():
     parser = argparse.ArgumentParser()
-    parser.add_argument("--task", type=str, default="FetchReach-v3")
+    parser.add_argument("--task", type=str, default="FetchReach-v2")
```
Comment (MischaPanch) [inline examples/mujoco/fetch_her_ddpg.py:34]: why this change?

File: examples/mujoco/fetch_her_ddpg.py [hunk 5]
```diff
@@ -96,6 +95,8 @@ def test_ddpg(args=get_args()):
 
     # logger
     if args.logger == "wandb":
+        import wandb
```
Comment (MischaPanch) [inline examples/mujoco/fetch_her_ddpg.py]: If you do it here, you need to do it everywhere. I thought wandb was not an optional dependency, am I wrong?

File: pyproject.toml [hunk 6]
```diff
@@ -27,8 +27,10 @@ exclude = ["test/*", "examples/*", "docs/*"]
 
 [tool.poetry.dependencies]
 python = "^3.11"
+docstring-parser = "^0.15"
```
Comment (MischaPanch) [inline pyproject.toml]: This and jsonargparse should probably be within its own group. Maybe becoming optional later or even now, what do you think @opcode81 ?

Tianshou doesn't need these dependencies, but they are needed to run the examples. I guess it's a bit weird to have to install an extra in order to run examples. On the other side, we should avoid adding any unnecessary requirements to the core package, python package conflicts can be a real PITA

File: pyproject.toml [hunk 7]
```diff
@@ -44,28 +46,43 @@ virtualenv = [
...
 
 [tool.poetry.extras]
-atari = ["atari_py", "opencv-python"]
-mujoco = ["mujoco_py"]
+atari = ["ale-py", "autorom", "opencv-python", "shimmy"]
+box2d = ["box2d-py", "pygame", "swig"]
+classic-control = ["pygame"]
+classic_control = ["pygame"]
+mujoco = ["mujoco", "imageio"]
+mujoco-py = ["mujoco-py", "cython"]
```
Comment (MischaPanch) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 8]
```diff
@@ -44,28 +46,43 @@ virtualenv = [
...
+shimmy = { version = ">=0.1.0,<1.0", optional = true }
+swig = { version = "4.*", optional = true }
+vizdoom = { version = "*", optional = true }
 
 [tool.poetry.extras]
-atari = ["atari_py", "opencv-python"]
-mujoco = ["mujoco_py"]
+atari = ["ale-py", "autorom", "opencv-python", "shimmy"]
+box2d = ["box2d-py", "pygame", "swig"]
+classic-control = ["pygame"]
```
Comment (MischaPanch) [inline pyproject.toml]: Is poetry not warning against duplicated entries?

```suggestion
```

File: pyproject.toml [hunk 9]
```diff
@@ -44,28 +46,43 @@ virtualenv = [
...
 pybullet = { version = "*", optional = true }
+pygame = { version = ">=2.1.3", optional = true }
+shimmy = { version = ">=0.1.0,<1.0", optional = true }
+swig = { version = "4.*", optional = true }
+vizdoom = { version = "*", optional = true }
 
 [tool.poetry.extras]
-atari = ["atari_py", "opencv-python"]
-mujoco = ["mujoco_py"]
+atari = ["ale-py", "autorom", "opencv-python", "shimmy"]
```
Comment (MischaPanch) [inline pyproject.toml:73]: Please explain in a block comment why we have to do it like that (with a link to the relevant poetry issue).

This part requires attention and monitoring gymnasium for changes

File: README.md [hunk 10]
```diff
@@ -96,6 +96,23 @@ You can also install the dev requirements by adding `--with dev` or the extras
...
+Available extras are:
+- `atari` (for Atari environments)
+- `box2d` (for Box2D environments)
+- `classic_control` (for classic control (discrete) environments)
+- `mujoco` (for MuJoCo environments)
+- `mujoco-py` (for legacy mujoco-py environments[^1])
+- `pybullet` (for pybullet environments)
+- `robotics` (for gymnasium-robotics environments)
+- `vizdoom` (for ViZDoom environments)
+- `envpool` (for envpool integration)
```
Comment (MischaPanch) [inline README.md]:
```suggestion
- `envpool` (for [envpool](https://github.com/sail-sg/envpool/)  integration)
```

**Review by MischaPanch — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (maxhuettenrauch) [reply README.md]: If you're referring to 'dev', then you're right. 'robotics' and 'vizdoom' should be extras (in the same fashion as 'atari' and 'mujoco' before). Should I just remove it from readme?

File: examples/box2d/bipedal_hardcore_sac.py (hunk 3 above)
Comment (maxhuettenrauch) [reply examples/box2d/bipedal_hardcore_sac.py]: I guess it was some older example that noone touched in a while. The check for old vs new api is still present in atari_wrapper.py, that's where I just copied it from. You're right though, the Collector assumes terminated and truncated coming from env.step, I'll just delete this part

File: examples/mujoco/fetch_her_ddpg.py (hunk 4 above)
Comment (maxhuettenrauch) [reply examples/mujoco/fetch_her_ddpg.py:34]: There is no v3 in gymnasium robotics (https://robotics.farama.org/envs/fetch/)

File: examples/mujoco/fetch_her_ddpg.py (hunk 5 above)
Comment (maxhuettenrauch) [reply examples/mujoco/fetch_her_ddpg.py]: I checked, it's the only example where wandb is import. I also thought about removing it since wandb currently is only part of the 'dev' group dependencies

File: README.md (hunk 2 above)
Comment (MischaPanch) [reply README.md]: Yes, just dev. Feel free to include it separately in the readme if not done yet, as groups are installed with a different command (not --extras)

File: examples/mujoco/fetch_her_ddpg.py (hunk 5 above)
Comment (MischaPanch) [reply examples/mujoco/fetch_her_ddpg.py]: Remove pls

File: pyproject.toml (hunk 6 above)
Comment (maxhuettenrauch) [reply pyproject.toml]: If we want it to be pip-installable, it should be an extra (for example 'high-level-api'). A group would only be installable using poetry

Comment (MischaPanch) [reply pyproject.toml]: I guess an extra called `argparse` would make sense. You'd need to very visibly document that this extra is required to run the high-level-api examples

**Review by MischaPanch — APPROVED** [review]

**Comment by MischaPanch** [comment]
Thank you, now only a final testing on windows and ubuntu is needed, and #1015 can be resolved. Merging this when the pipeline passed
